File: packages/clinical-contracts/src/rule4/registryMerge.ts
export type Rule4RegistryEntry = {
  code: string;
  namespace: 'reason' | 'limitation';
  source: string;
};

export class Rule4RegistryMergeError extends Error {
  readonly code: string;

  constructor(code: string) {
    super(code);
    this.code = code;
  }
}

export function mergeRule4RegistryEntries(
  phase1: readonly Rule4RegistryEntry[],
  phase2: readonly Rule4RegistryEntry[],
): Rule4RegistryEntry[] {
  const merged = new Map<string, Rule4RegistryEntry>();
  for (const entry of [...phase1, ...phase2]) {
    const existing = merged.get(entry.code);
    if (!existing) {
      merged.set(entry.code, entry);
      continue;
    }
    if (existing.namespace !== entry.namespace) {
      throw new Rule4RegistryMergeError('RULE4_REGISTRY_MERGE_NAMESPACE_CONFLICT');
    }
    if (existing.source !== entry.source) {
      throw new Rule4RegistryMergeError('RULE4_REGISTRY_MERGE_SOURCE_CONFLICT');
    }
  }
  return [...merged.values()];
}

File: apps/clinical-engine/src/ehas2_clinical_engine/rule4/registry_merge.py
from __future__ import annotations


class Rule4RegistryMergeError(ValueError):
    code = "RULE4_REGISTRY_MERGE_CONFLICT"


def merge_registry_entries(phase1_entries: list[dict], phase2_entries: list[dict]) -> list[dict]:
    merged: dict[str, dict] = {}
    for entry in phase1_entries + phase2_entries:
        code = entry["code"]
        namespace = entry.get("namespace")
        source = entry.get("source")
        existing = merged.get(code)
        if existing is None:
            merged[code] = entry
            continue
        if existing.get("namespace") != namespace:
            raise Rule4RegistryMergeError("RULE4_REGISTRY_MERGE_NAMESPACE_CONFLICT")
        if existing.get("source") != source:
            raise Rule4RegistryMergeError("RULE4_REGISTRY_MERGE_SOURCE_CONFLICT")
    return list(merged.values())
